File: problem_016.c
// -*- compile-command: "gcc -std=c99 -o problem_016 problem_016.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 16:\n"
           "What is the sum of the digits of the number 2^1000\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    int bignum[350] = {0};
    int total = 0;
    bool debug = false;

    bignum[1] = 2;

    for (int i=1 ; i<1000 ; i++)
    {
        //Multiply each element
        for (int n=0 ; n<350 ; n++)
            bignum[n]=bignum[n]*2;

        //Spill each element if needed
        for (int n=0 ; n<350 ; n++)
        {
            while (bignum[n] > 9)
            {
                bignum[n+1]=bignum[n+1]+1;
                bignum[n]=bignum[n]-10;
            }
        }

        // DEBUG: Print each element
        if (debug == true)
        {
            for (int n=350 ; n>1 ; n--)
                printf("%i",bignum[n]);
            printf("\n\n");
        }
    }

    // Add digits together
    for (int n=0 ; n<350 ; n++)
        total += bignum[n];

    printf("Total: %i\n", total);
    printElapsedTime(start);

    return 0;
}

File: problem_016.go
// -*- compile-command: "go build -o problem_016_go problem_016.go" -*-
// Copyright (c) 2017 Michael Caldwell
package main

import "fmt"

func main() {
        fmt.Println("Project Euler - Problem 16:")
        fmt.Println("What is the sum of the digits of the number 2^1000\n")

	bigNumber := make([]int, 500)
	bigNumber[0] = 2
	debug := false

	for n := 1; n < 1000; n++ {
		// Multiply
		for i := 0; i < len(bigNumber); i++ {
			bigNumber[i] *= 2
		}

		// Carry
		for i := 0; i < len(bigNumber) - 1; i++ {
			for bigNumber[i] > 9 {
				bigNumber[i] -= 10
				bigNumber[i+1]++
			}
		}
	}

	// Print the big number
	if debug == true {
		print := false
		for i := len(bigNumber)-1; i >= 0; i-- {
			if bigNumber[i] != 0 { print = true }
			if print == true { fmt.Print(bigNumber[i]) }
		}
		fmt.Println("\n")
	}

	sum := 0
	// Add
	for i := 0; i < len(bigNumber); i++ {
		sum += bigNumber[i]
	}

        fmt.Println(sum)
}
